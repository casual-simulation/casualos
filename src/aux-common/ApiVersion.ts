/**
 * Gets the API version that the AUX common library currently supports.
 * This needs to get manually incremented whenever a new API version is added.
 * Basically a really barebones start for API versioning.
 */
export const apiVersion = 2;
