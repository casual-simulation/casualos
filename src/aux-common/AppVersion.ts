/* CasualOS is a set of web-based tools designed to facilitate the creation of real-time, multi-user, context-aware interactive experiences.
 *
 * Copyright (c) 2019-2025 Casual Simulation, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

export interface AppVersion {
    /**
     * Gets the git tag that this version was built from.
     */
    gitTag: string;

    /**
     * Gets the Git hash that this version was built from.
     */
    gitHash: string;

    /**
     * Gets the API that this app has support for.
     */
    apiVersion: number;
}
