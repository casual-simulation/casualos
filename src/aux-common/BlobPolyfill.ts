if (!globalThis.Blob) {
    if (typeof process === 'undefined') {
        console.warn(
            '[BlobPolyfill] Loading Blob stub! This should not happen in browser environments!'
        );
    }
    (<any>globalThis).Blob = class {
        type: string;
        parts: any[];

        constructor(parts: any[], options: any) {
            this.parts = parts.slice();
            this.type = options?.type;
        }

        async arrayBuffer(): Promise<ArrayBuffer> {
            let bytes = [] as number[];
            const encoder = new TextEncoder();
            for (const part of this.parts) {
                if (typeof part === 'string') {
                    const encoded = encoder.encode(part);
                    bytes.push(...encoded);
                } else if (part instanceof ArrayBuffer) {
                    bytes.push(...new Uint8Array(part));
                } else if (ArrayBuffer.isView(part)) {
                    bytes.push(...new Uint8Array(part.buffer));
                } else if (part instanceof Blob) {
                    const buffer = await part.arrayBuffer();
                    bytes.push(...new Uint8Array(buffer));
                }
            }

            const view = new Uint8Array(bytes);
            return view.buffer;
        }
    };
}
