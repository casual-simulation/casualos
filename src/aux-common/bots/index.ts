export * from './Bot';
export * from './BotEvents';
export * from './BotCalculationContext';
export * from './BotCalculationContextFactory';
export * from './BotCalculations';
export * from './BotActions';
export * from './TagUpdatedEvent';
export * from './BotIndex';
export * from './BotLookupTable';
export * from './BotLookupTableHelper';
export * from './StateUpdatedEvent';
export * from './StoredAux';
export * from './AuxStateHelpers';
