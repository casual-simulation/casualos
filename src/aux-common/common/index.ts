export * from './Action';
export * from './ConnectionInfo';
export * from './RemoteActions';
export * from './ConnectionIndicator';
export * from './StatusUpdate';
export * from './CurrentVersion';
export * from './StatusUpdateUtils';
export * from './LoadingProgress';
export * from './ConnectionToken';
export * from './Iterators';
export * from './PolicyPermissions';
export * from './DenialReason';
