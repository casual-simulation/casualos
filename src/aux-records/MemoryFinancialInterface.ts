/* CasualOS is a set of web-based tools designed to facilitate the creation of real-time, multi-user, context-aware interactive experiences.
 *
 * Copyright (c) 2019-2025 Casual Simulation, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import { cloneDeep } from 'lodash';
import type { FinancialInterface } from './financial/FinancialInterface';
import { AccountCodes } from './financial/FinancialInterface';
import type {
    Account,
    AccountBalance,
    AccountFilter,
    CreateAccountsError,
    CreateTransfersError,
    QueryFilter,
    Transfer,
} from './financial/Types';
import {
    AccountFlags,
    CreateAccountError,
    CreateTransferError,
    TransferFlags,
} from './financial/Types';

/**
 * The max bigint tigerbeetle can handle. 2^128 - 1 is reserved.
 */
const MAX_BIGINT_128 = BigInt('340282366920938463463374607431768211454');

export class MemoryFinancialInterface implements FinancialInterface {
    private _accounts: Map<Account['id'], Account> = new Map();
    private _transfers: Transfer[] = [];
    private _currentId = 0n;

    private _validateAccount(account: Account): CreateAccountError {
        if (this._accounts.has(account.id)) {
            return CreateAccountError.exists;
        } else if (account.id === 0n) {
            return CreateAccountError.id_must_not_be_zero;
        } else if (account.code > MAX_BIGINT_128) {
            return CreateAccountError.id_must_not_be_int_max;
        } else if (
            account.timestamp !== 0n &&
            account.flags & AccountFlags.imported
        ) {
            return CreateAccountError.timestamp_must_be_zero;
        } else {
            return CreateAccountError.ok;
        }

        // TODO: Add more validation
    }

    private _validateTransfer(transfer: Transfer): CreateTransferError {
        /**
         * * Format and convention validation
         */
        if (transfer.id === 0n) {
            return CreateTransferError.id_must_not_be_zero;
        }
        if (transfer.id > MAX_BIGINT_128) {
            return CreateTransferError.id_must_not_be_int_max;
        }
        const existingTransfer = this._transfers.find(
            (t) => t.id === transfer.id
        );
        if (existingTransfer) {
            if (existingTransfer.flags !== transfer.flags) {
                return CreateTransferError.exists_with_different_flags;
            }
            if (existingTransfer.pending_id !== transfer.pending_id) {
                return CreateTransferError.exists_with_different_pending_id;
            }
            if (existingTransfer.timeout !== transfer.timeout) {
                return CreateTransferError.exists_with_different_timeout;
            }
            if (
                existingTransfer.debit_account_id !== transfer.debit_account_id
            ) {
                return CreateTransferError.exists_with_different_debit_account_id;
            }
            if (
                existingTransfer.credit_account_id !==
                transfer.credit_account_id
            ) {
                return CreateTransferError.exists_with_different_credit_account_id;
            }
            if (existingTransfer.user_data_128 !== transfer.user_data_128) {
                return CreateTransferError.exists_with_different_user_data_128;
            }
            if (existingTransfer.user_data_64 !== transfer.user_data_64) {
                return CreateTransferError.exists_with_different_user_data_64;
            }
            if (existingTransfer.user_data_32 !== transfer.user_data_32) {
                return CreateTransferError.exists_with_different_user_data_32;
            }
            if (existingTransfer.ledger !== transfer.ledger) {
                return CreateTransferError.exists_with_different_ledger;
            }
            if (existingTransfer.code !== transfer.code) {
                return CreateTransferError.exists_with_different_code;
            }
            if (existingTransfer.amount !== transfer.amount) {
                if (existingTransfer.flags & TransferFlags.balancing_debit) {
                    // TODO: Implement balancing debit logic
                } else if (
                    existingTransfer.flags & TransferFlags.balancing_credit
                ) {
                    // TODO: Implement balancing credit logic
                } else if (
                    existingTransfer.flags & TransferFlags.post_pending_transfer
                ) {
                    // TODO: Implement post pending transfer logic
                } else {
                    return CreateTransferError.exists_with_different_amount;
                }
            }
        }

        /**
         * * Existential validation
         * 1. Credit account must exist
         * 2. Debit account must exist
         */
        let creditAccount = this._accounts.get(transfer.credit_account_id);
        let debitAccount = this._accounts.get(transfer.debit_account_id);
        if (
            transfer.flags & TransferFlags.void_pending_transfer ||
            transfer.flags & TransferFlags.post_pending_transfer
        ) {
            if (!transfer.pending_id || transfer.pending_id === 0n) {
                return CreateTransferError.pending_id_must_not_be_zero;
            }
            if (transfer.pending_id > MAX_BIGINT_128) {
                return CreateTransferError.pending_id_must_not_be_int_max;
            }
            const pendingTransfer =
                this._transfers[
                    this._transfers.findIndex(
                        (t) => t.id === transfer.pending_id
                    )
                ];
            if (!pendingTransfer) {
                return CreateTransferError.pending_transfer_not_found;
            }
            if (pendingTransfer.flags & TransferFlags.pending) {
                return CreateTransferError.pending_transfer_not_pending;
            }
            if (pendingTransfer.flags & TransferFlags.void_pending_transfer) {
                return CreateTransferError.pending_transfer_already_voided;
            }
            if (pendingTransfer.flags & TransferFlags.post_pending_transfer) {
                return CreateTransferError.pending_transfer_already_posted;
            }

            creditAccount = this._accounts.get(
                pendingTransfer.credit_account_id
            );
            debitAccount = this._accounts.get(pendingTransfer.debit_account_id);
        }
        if (!creditAccount) {
            return CreateTransferError.credit_account_not_found;
        }
        if (!debitAccount) {
            return CreateTransferError.debit_account_not_found;
        }

        /**
         * TODO: Add more validation
         * * Ensure the transfer satisfies the account codes
         * * Ensure the transfer satisfies the flags
         */
    }

    private _performTransfer(transfer: Transfer): CreateTransferError {
        const validation = this._validateTransfer(transfer);
        if (validation !== CreateTransferError.ok) return validation;
        let creditAccount = this._accounts.get(transfer.credit_account_id);
        let debitAccount = this._accounts.get(transfer.debit_account_id);
        if (transfer.flags & TransferFlags.pending) {
            creditAccount.credits_pending += transfer.amount;
            debitAccount.debits_pending += transfer.amount;
        } else if (transfer.flags & TransferFlags.void_pending_transfer) {
            if (!creditAccount || !debitAccount) {
                const voidTransfer =
                    this._transfers[
                        this._transfers.findIndex(
                            (t) => t.id === transfer.pending_id
                        )
                    ];
                creditAccount = this._accounts.get(
                    voidTransfer.credit_account_id
                );
                debitAccount = this._accounts.get(
                    voidTransfer.debit_account_id
                );
            }
            creditAccount.credits_pending -= transfer.amount;
            debitAccount.debits_pending -= transfer.amount;
        } else if (transfer.flags & TransferFlags.post_pending_transfer) {
            if (!creditAccount || !debitAccount) {
                const postTransfer =
                    this._transfers[
                        this._transfers.findIndex(
                            (t) => t.id === transfer.pending_id
                        )
                    ];
                creditAccount = this._accounts.get(
                    postTransfer.credit_account_id
                );
                debitAccount = this._accounts.get(
                    postTransfer.debit_account_id
                );
            }
            creditAccount.credits_pending -= transfer.amount;
            debitAccount.debits_pending -= transfer.amount;
            creditAccount.credits_posted += transfer.amount;
            debitAccount.debits_posted += transfer.amount;
        } else {
            creditAccount.credits_posted += transfer.amount;
            debitAccount.debits_posted += transfer.amount;
        }
        /**
         * TODO: Continue transfer logic implementation
         * * 1. Route based on if the credits and debits need to be pending or posted
         * * 2. Ensure the transfer satisfies the accounting equation
         */
    }

    generateId = () => {
        return this._currentId++;
    };

    async createAccount(account: Account): Promise<CreateAccountsError[]> {
        const errs = [{ index: 0, result: this._validateAccount(account) }];
        if (errs[0].result === CreateAccountError.ok)
            this._accounts.set(account.id, account);
        return errs;
    }

    async createAccounts(batch: Account[]): Promise<CreateAccountsError[]> {
        return batch.map((account, index) => {
            const error = this._validateAccount(account);
            if (error === CreateAccountError.ok) {
                this._accounts.set(account.id, account);
            }
            return { index, result: error };
        });
    }

    async createTransfers(batch: Transfer[]): Promise<CreateTransfersError[]> {
        const errs: CreateTransfersError[] = [];
        for (let i = 0; i < batch.length; i++) {
            errs.push({
                index: i,
                result: this._validateTransfer(batch[i]),
            });
            if (errs[i].result === CreateTransferError.ok) {
                this._transfers.push(batch[i]);
            }
        }
        return errs;
    }

    async lookupAccounts(batch: Account['id'][]): Promise<Account[]> {
        const accounts: Account[] = [];
        for (const id of batch) {
            const account = this._accounts.get(id);
            if (account) {
                accounts.push(cloneDeep(account));
            }
        }
        return accounts;
    }

    lookupTransfers(batch: Transfer['id'][]): Promise<Transfer[]> {
        throw new Error('Method not implemented.');
    }

    getAccountTransfers(filter: AccountFilter): Promise<Transfer[]> {
        throw new Error('Method not implemented.');
    }

    getAccountBalances(filter: AccountFilter): Promise<AccountBalance[]> {
        throw new Error('Method not implemented.');
    }

    queryAccounts(filter: QueryFilter): Promise<Account[]> {
        throw new Error('Method not implemented.');
    }

    queryTransfers(filter: QueryFilter): Promise<Transfer[]> {
        throw new Error('Method not implemented.');
    }
}
