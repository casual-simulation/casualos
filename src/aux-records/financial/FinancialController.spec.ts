/* CasualOS is a set of web-based tools designed to facilitate the creation of real-time, multi-user, context-aware interactive experiences.
 *
 * Copyright (c) 2019-2025 Casual Simulation, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { failure, success, unwrap } from '@casual-simulation/aux-common';
import {
    FinancialController,
    getAccountBalance,
    getAssetAccountBalance,
    getLiabilityAccountBalance,
} from './FinancialController';
import {
    ACCOUNT_IDS,
    AccountCodes,
    LEDGERS,
    TransferCodes,
} from './FinancialInterface';
import { MemoryFinancialInterface } from './MemoryFinancialInterface';
import type { Account } from './Types';
import { AccountFlags, TransferFlags } from './Types';

console.error = jest.fn();

describe('FinancialController', () => {
    let financialInterface: MemoryFinancialInterface;
    let controller: FinancialController;
    let dateNowMock: jest.Mock<number>;

    const realDateNow = Date.now;

    beforeEach(() => {
        dateNowMock = Date.now = jest.fn(() => 123);
        financialInterface = new MemoryFinancialInterface();
        controller = new FinancialController(financialInterface);
    });

    afterEach(() => {
        Date.now = realDateNow;
    });

    describe('init()', () => {
        it('should create a stripe assets account', async () => {
            await controller.init();

            expect([...financialInterface.accounts.values()]).toEqual([
                {
                    id: 1001n,
                    debits_pending: 0n,
                    debits_posted: 0n,
                    credits_pending: 0n,
                    credits_posted: 0n,
                    user_data_128: 0n,
                    user_data_64: 0n,
                    user_data_32: 0,
                    reserved: 0,
                    ledger: 1,
                    flags: AccountFlags.credits_must_not_exceed_debits,
                    code: AccountCodes.assets_cash,
                    timestamp: 0n,
                },
            ]);
        });
    });

    describe('createAccount()', () => {
        it('should create a new account with the specified code', async () => {
            const account = await controller.createAccount(
                AccountCodes.assets_cash
            );
            expect(account).toEqual(
                success({
                    id: '1',
                })
            );

            expect([...financialInterface.accounts.values()]).toEqual([
                {
                    id: 1n,
                    debits_pending: 0n,
                    debits_posted: 0n,
                    credits_pending: 0n,
                    credits_posted: 0n,
                    user_data_128: 0n,
                    user_data_64: 0n,
                    user_data_32: 0,
                    reserved: 0,
                    ledger: 1,
                    flags: AccountFlags.credits_must_not_exceed_debits,
                    code: AccountCodes.assets_cash,
                    timestamp: 0n,
                },
            ]);
        });

        it('should be able to create a user account', async () => {
            const account = await controller.createAccount(
                AccountCodes.liabilities_user
            );
            expect(account).toEqual(
                success({
                    id: '1',
                })
            );

            expect([...financialInterface.accounts.values()]).toEqual([
                {
                    id: 1n,
                    debits_pending: 0n,
                    debits_posted: 0n,
                    credits_pending: 0n,
                    credits_posted: 0n,
                    user_data_128: 0n,
                    user_data_64: 0n,
                    user_data_32: 0,
                    reserved: 0,
                    ledger: 1,
                    flags: AccountFlags.debits_must_not_exceed_credits,
                    code: AccountCodes.liabilities_user,
                    timestamp: 0n,
                },
            ]);
        });
    });

    describe('getAccount()', () => {
        beforeEach(async () => {
            await controller.init();
            await controller.createAccount(AccountCodes.assets_cash);
            await controller.createAccount(AccountCodes.liabilities_user);
        });

        it('should return the account with the specified ID', async () => {
            const account = await controller.getAccount(1n);
            expect(account).toEqual(
                success({
                    id: 1n,
                    debits_pending: 0n,
                    debits_posted: 0n,
                    credits_pending: 0n,
                    credits_posted: 0n,
                    user_data_128: 0n,
                    user_data_64: 0n,
                    user_data_32: 0,
                    reserved: 0,
                    ledger: 1,
                    flags: AccountFlags.credits_must_not_exceed_debits,
                    code: AccountCodes.assets_cash,
                    timestamp: 0n,
                })
            );
        });
    });

    describe('internalTransfer()', () => {
        let account1Id: string;
        let account2Id: string;

        beforeEach(async () => {
            unwrap(await controller.init());

            ({ id: account1Id } = unwrap(
                await controller.createAccount(AccountCodes.liabilities_user)
            ));
            ({ id: account2Id } = unwrap(
                await controller.createAccount(AccountCodes.liabilities_user)
            ));
        });

        it('should be able to transfer money from the assets_cash account to a user account', async () => {
            const result = await controller.internalTransfer({
                transfers: [
                    {
                        debitAccountId: ACCOUNT_IDS.stripe_assets,
                        creditAccountId: account1Id,
                        currency: 'credits',
                        amount: 100n,
                        code: TransferCodes.admin_credit,
                    },
                ],
            });

            expect(result).toEqual(
                success({
                    transactionId: '3',
                    transferIds: ['4'],
                })
            );
            expect(financialInterface.transfers).toEqual([
                {
                    id: 4n,
                    amount: 100n,
                    code: TransferCodes.admin_credit,
                    credit_account_id: BigInt(account1Id),
                    debit_account_id: ACCOUNT_IDS.stripe_assets,
                    flags: TransferFlags.none,
                    ledger: LEDGERS.credits,
                    pending_id: 0n,
                    timeout: 0,
                    timestamp: 0n,
                    user_data_128: 3n,
                    user_data_64: 0n,
                    user_data_32: 0,
                },
            ]);
        });

        it('should be able to transfer money from one user account to another', async () => {
            unwrap(
                await controller.internalTransfer({
                    transfers: [
                        {
                            debitAccountId: ACCOUNT_IDS.stripe_assets,
                            creditAccountId: account1Id,
                            currency: 'credits',
                            amount: 1000n,
                            code: TransferCodes.admin_credit,
                        },
                    ],
                })
            );

            const result = await controller.internalTransfer({
                transfers: [
                    {
                        debitAccountId: account1Id,
                        creditAccountId: account2Id,
                        currency: 'credits',
                        amount: 100n,
                        code: TransferCodes.admin_credit,
                    },
                ],
            });

            expect(result).toEqual(
                success({
                    transactionId: '5',
                    transferIds: ['6'],
                    // accountBalances: [
                    //     {
                    //         accountId: account1Id,
                    //         balance: 900n,
                    //     },
                    //     {
                    //         accountId: account2Id,
                    //         balance: 100n,
                    //     },
                    // ],
                })
            );
            expect(financialInterface.transfers.slice(1)).toEqual([
                {
                    id: 6n,
                    amount: 100n,
                    code: TransferCodes.admin_credit,
                    credit_account_id: BigInt(account2Id),
                    debit_account_id: BigInt(account1Id),
                    flags: TransferFlags.none,
                    ledger: LEDGERS.credits,
                    pending_id: 0n,
                    timeout: 0,
                    timestamp: 0n,
                    user_data_128: 5n,
                    user_data_64: 0n,
                    user_data_32: 0,
                },
            ]);
        });

        it('should use the given transfer Id', async () => {
            const result = await controller.internalTransfer({
                transfers: [
                    {
                        transferId: 100n,
                        debitAccountId: ACCOUNT_IDS.stripe_assets,
                        creditAccountId: account1Id,
                        currency: 'credits',
                        amount: 100n,
                        code: TransferCodes.admin_credit,
                    },
                ],
            });

            expect(result).toEqual(
                success({
                    transactionId: '3',
                    transferIds: ['100'],
                    // accountBalances: [
                    //     {
                    //         accountId: ACCOUNT_IDS.stripe_assets,
                    //         balance: -100n,
                    //     },
                    //     {
                    //         accountId: account1Id,
                    //         balance: 100n,
                    //     },
                    // ],
                })
            );
            expect(financialInterface.transfers).toEqual([
                {
                    id: 100n,
                    amount: 100n,
                    credit_account_id: BigInt(account1Id),
                    debit_account_id: ACCOUNT_IDS.stripe_assets,
                    code: TransferCodes.admin_credit,
                    flags: TransferFlags.none,
                    ledger: LEDGERS.credits,
                    pending_id: 0n,
                    timeout: 0,
                    timestamp: 0n,
                    user_data_128: 3n,
                    user_data_64: 0n,
                    user_data_32: 0,
                },
            ]);
        });

        it('should reject the transfer if it would cause a user account to go negative', async () => {
            const result = await controller.internalTransfer({
                transfers: [
                    {
                        transferId: 100n,
                        debitAccountId: account1Id,
                        creditAccountId: account2Id,
                        currency: 'credits',
                        amount: 100n,
                        code: TransferCodes.admin_credit,
                    },
                ],
            });

            expect(result).toEqual(
                failure({
                    errorCode: 'debits_exceed_credits',
                    errorMessage:
                        'The transfer would cause the account debits to exceed its credits.',
                    accountId: account1Id,
                })
            );
            expect(financialInterface.transfers).toEqual([]);
        });

        it('should be able to perform transfers in a transaction', async () => {
            const result = await controller.internalTransfer({
                transfers: [
                    {
                        transferId: 100n,
                        debitAccountId: ACCOUNT_IDS.stripe_assets,
                        creditAccountId: account1Id,
                        currency: 'credits',
                        amount: 100n,
                        code: TransferCodes.admin_credit,
                    },
                    {
                        transferId: 101n,
                        debitAccountId: account1Id,
                        creditAccountId: account2Id,
                        currency: 'credits',
                        amount: 100n,
                        code: TransferCodes.admin_credit,
                    },
                ],
            });

            expect(result).toEqual(
                success({
                    transactionId: '3',
                    transferIds: ['100', '101'],
                    // accountBalances: [
                    //     {
                    //         accountId: ACCOUNT_IDS.stripe_assets,
                    //         balance: -100n,
                    //     },
                    //     {
                    //         accountId: account1Id,
                    //         balance: 0n,
                    //     },
                    //     {
                    //         accountId: account2Id,
                    //         balance: 100n,
                    //     },
                    // ],
                })
            );

            expect(financialInterface.transfers).toEqual([
                {
                    id: 100n,
                    amount: 100n,
                    credit_account_id: BigInt(account1Id),
                    debit_account_id: ACCOUNT_IDS.stripe_assets,
                    code: TransferCodes.admin_credit,
                    flags: TransferFlags.linked,
                    ledger: LEDGERS.credits,
                    pending_id: 0n,
                    timeout: 0,
                    timestamp: 0n,

                    // should put the transaction id in user_data_128
                    user_data_128: 3n,
                    user_data_64: 0n,
                    user_data_32: 0,
                },
                {
                    id: 101n,
                    amount: 100n,
                    credit_account_id: BigInt(account2Id),
                    debit_account_id: BigInt(account1Id),
                    code: TransferCodes.admin_credit,
                    flags: TransferFlags.none,
                    ledger: LEDGERS.credits,
                    pending_id: 0n,
                    timeout: 0,
                    timestamp: 0n,

                    // should put the transaction id in user_data_128
                    user_data_128: 3n,
                    user_data_64: 0n,
                    user_data_32: 0,
                },
            ]);
        });
    });
});

describe('getAccountBalance()', () => {
    it('should return the user balance for liabilities_user accounts', async () => {
        const account: Account = {
            id: 1n,
            debits_pending: 100n,
            debits_posted: 50n,
            credits_pending: 200n,
            credits_posted: 155n,
            user_data_128: 0n,
            user_data_64: 0n,
            user_data_32: 0,
            reserved: 0,
            ledger: 1,
            flags: AccountFlags.debits_must_not_exceed_credits,
            code: AccountCodes.liabilities_user,
            timestamp: 0n,
        };

        const balance = getAccountBalance(account);

        expect(balance).toBe(105); // 155 - 50 = 105
    });

    it('should return the assets balance for assets_cash accounts', async () => {
        const account: Account = {
            id: 1n,
            debits_pending: 200n,
            debits_posted: 155n,
            credits_pending: 100n,
            credits_posted: 50n,
            user_data_128: 0n,
            user_data_64: 0n,
            user_data_32: 0,
            reserved: 0,
            ledger: 1,
            flags: AccountFlags.debits_must_not_exceed_credits,
            code: AccountCodes.assets_cash,
            timestamp: 0n,
        };

        const balance = getAccountBalance(account);

        expect(balance).toBe(105); // 155 - 50 = 105
    });
});

describe('getLiabilityAccountBalance()', () => {
    it('should return the posted credits minus the debits', async () => {
        const account: Account = {
            id: 1n,
            debits_pending: 100n,
            debits_posted: 50n,
            credits_pending: 200n,
            credits_posted: 155n,
            user_data_128: 0n,
            user_data_64: 0n,
            user_data_32: 0,
            reserved: 0,
            ledger: 1,
            flags: AccountFlags.debits_must_not_exceed_credits,
            code: AccountCodes.liabilities_user,
            timestamp: 0n,
        };

        const balance = getLiabilityAccountBalance(account);

        expect(balance).toBe(105); // 155 - 50 = 105
    });
});

describe('getAssetAccountBalance()', () => {
    it('should return the posted debits minus the credits', async () => {
        const account: Account = {
            id: 1n,
            debits_pending: 200n,
            debits_posted: 155n,
            credits_pending: 100n,
            credits_posted: 50n,
            user_data_128: 0n,
            user_data_64: 0n,
            user_data_32: 0,
            reserved: 0,
            ledger: 1,
            flags: AccountFlags.credits_must_not_exceed_debits,
            code: AccountCodes.assets_cash,
            timestamp: 0n,
        };

        const balance = getAssetAccountBalance(account);

        expect(balance).toBe(105); // 155 - 50 = 105
    });
});
