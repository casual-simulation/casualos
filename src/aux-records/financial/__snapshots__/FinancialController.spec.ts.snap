// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinancialController init() should create all the default accounts 1`] = `
Array [
  Object {
    "code": 1001,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 12,
    "id": 1001,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1001,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 12,
    "id": 1002,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 4101,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 10,
    "id": 4101,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 4101,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 10,
    "id": 4102,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 6001,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 2,
    "id": 6001,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 6001,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 4,
    "id": 6002,
    "ledger": 2,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 0,
    "id": 9999,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 4,
    "id": 9997,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 2,
    "id": 9998,
    "ledger": 1,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 0,
    "id": 9996,
    "ledger": 2,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 4,
    "id": 9994,
    "ledger": 2,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
  Object {
    "code": 1,
    "credits_pending": 0,
    "credits_posted": 0,
    "debits_pending": 0,
    "debits_posted": 0,
    "flags": 2,
    "id": 9995,
    "ledger": 2,
    "reserved": 0,
    "user_data_128": 0,
    "user_data_32": 0,
    "user_data_64": 0,
  },
]
`;
